Firestore/core/src/firebase/firestore/local/memory_persistence.h

```diff
@@ -0,0 +1,125 @@
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_MEMORY_PERSISTENCE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_MEMORY_PERSISTENCE_H_
+
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/auth/user.h"
+#include "Firestore/core/src/firebase/firestore/local/memory_index_manager.h"
+#include "Firestore/core/src/firebase/firestore/local/memory_mutation_queue.h"
+#include "Firestore/core/src/firebase/firestore/local/memory_query_cache.h"
+#include "Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.h"
+#include "Firestore/core/src/firebase/firestore/local/persistence.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+struct LruParams;
+class MemoryIndexManager;
+class MemoryMutationQueue;
+class MemoryQueryCache;
+class MemoryRemoteDocumentCache;
+class MutationQueue;
+class QueryCache;
+class ReferenceDelegate;
+class RemoteDocumentCache;
+class Sizer;
+
+/**
+ * An in-memory implementation of the Persistence interface. Values are stored
+ * only in RAM and are never persisted to any durable storage.
+ */
+class MemoryPersistence : public Persistence {
+ public:
+  using MutationQueues =
+      std::unordered_map<auth::User,
+                         std::unique_ptr<MemoryMutationQueue>,
+                         auth::HashUser>;
+
+  static std::unique_ptr<MemoryPersistence> WithEagerGarbageCollector();
+
+  static std::unique_ptr<MemoryPersistence> WithLruGarbageCollector(
+      LruParams params, std::unique_ptr<Sizer> sizer);
+
+  const MutationQueues& mutation_queues() {
+    return mutation_queues_;
+  }
+
+  // MARK: Persistence overrides
+
+  model::ListenSequenceNumber current_sequence_number() override;
+
+  void Shutdown() override;
+
+  MemoryMutationQueue* GetMutationQueueForUser(const auth::User& user) override;
+
+  MemoryQueryCache* query_cache() override;
+
+  MemoryRemoteDocumentCache* remote_document_cache() override;
+
+  MemoryIndexManager* index_manager() override;
+
+  ReferenceDelegate* reference_delegate() override;
+
+ protected:
+  void RunInternal(absl::string_view label,
+                   std::function<void()> block) override;
+
+ private:
+  MemoryPersistence();
+
+  void set_reference_delegate(std::unique_ptr<ReferenceDelegate> delegate) {
+    reference_delegate_ = std::move(delegate);
+  }
+
+  bool started_ = false;
+
+  MutationQueues mutation_queues_;
+
+  /**
+   * The QueryCache representing the persisted cache of queries.
+   *
+   * Note that this is retained here to make it easier to write tests affecting
+   * both the in-memory and LevelDB-backed persistence layers. Tests can create
+   * a new FSTLocalStore wrapping this Persistence instance and this will make
+   * the in-memory persistence layer behave as if it were actually persisting
+   * values.
+   */
+  std::unique_ptr<MemoryQueryCache> query_cache_;
```

Comment:
Question: do these have to be `unique_ptr`s? It seems like plain variables should work.